src/Microsoft.Identity.Client/MsalException.cs

```diff
@@ -1,122 +1,188 @@
+        public MsalException(string errorCode, string errorMessage)
+            : base(errorMessage)
+        {
+            ErrorCode = errorCode;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the exception class with a specified
+        /// error code and a reference to the inner exception that is the cause of
+        /// this exception.
+        /// </summary>
+        /// <param name="errorCode">
+        /// The error code returned by the service or generated by client. This is the code you can rely on
+        /// for exception handling.
+        /// </param>
+        /// <param name="errorMessage">The error message that explains the reason for the exception.</param>
+        /// <param name="innerException">
+        /// The exception that is the cause of the current exception, or a null reference if no inner
+        /// exception is specified.
+        /// </param>
+        public MsalException(string errorCode, string errorMessage, Exception innerException)
+            : base(errorMessage, innerException)
+        {
+            ErrorCode = errorCode;
+        }
+
+        /// <summary>
+        /// Gets the protocol error code returned by the service or generated by client. This is the code you can rely on for
+        /// exception handling. Values for this code are typically provided in constant strings in the derived exceptions types
+        /// with explanations of mitigation.
+        /// </summary>
+        public string ErrorCode { get; private set; }
+
+        /// <summary>
+        /// Creates and returns a string representation of the current exception.
+        /// </summary>
+        /// <returns>A string representation of the current exception.</returns>
+        public override string ToString()
+        {
+            return base.ToString() + string.Format(CultureInfo.InvariantCulture, "\n\tErrorCode: {0}", ErrorCode);
+        }
+
+        #region SERIALIZATION
+
+        private const string ExceptionTypeKey = "type";
+        private const string ErrorCodeKey = "error_code";
+        private const string ErrorDescriptionKey = "error_description";
+
+        internal virtual void PopulateJson(JObject jobj)
+        {
+            jobj[ExceptionTypeKey] = GetType().Name;
+        }
+
+        internal virtual void PopulateObjectFromJson(JObject jobj)
+        {
+        }
+
+        /// <summary>
+        /// Allows serialization of most values of the exception into JSON.
+        /// </summary>
+        /// <returns></returns>
+        public string ToJsonString()
+        {
+            JObject jobj = new JObject();
+            PopulateJson(jobj);
+            return jobj.ToString();
+        }
+
+        private static readonly Lazy<Dictionary<string, Type>> s_typeNameToType = new Lazy<Dictionary<string, Type>>(() =>
+        {
+            return new Dictionary<string, Type>
+            {
+                { typeof(MsalException).Name, typeof(MsalException) },
```

Comment:
What would happen if we or customers would derive a new type from MsalException?

Reply:
We wouldn't find it here. However not sure I understand the usecase of a customer creating their own MsalException? Nor do I see us creating a new exception, it should likely be possible to detect that by a utest if we wanted to.

---
In reply to: [264489006](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/967#discussion_r264489006) [](ancestors = 264489006)